src/unix/futex.c

```diff
@@ -294,3 +294,103 @@ init_futices(process p)
         halt("failed to allocate futex table\n");
 
 }
+
+/* robust mutex handling */
+
+#define FUTEX_OWNER_DIED	0x40000000
+#define FUTEX_KEY_ADDR(x, o)    ((int *)((u8 *)(x) + (o)))
+
+typedef struct robust_list {
+    struct robust_list *next;
+} *robust_list;
+
+typedef struct robust_list_head {
+    robust_list list;
+    long futex_offset;
+    void *list_op_pending;
+} *robust_list_head;
+
+static boolean valid_user_address(process p, void *a, int len)
+{
+    u64 pg = u64_from_pointer(a) & ~PAGEMASK;
+    u64 end = (u64_from_pointer(a) + len) & ~PAGEMASK;
+
+    for (; pg <= end; pg += PAGESIZE) {
+        if (vmap_from_vaddr(p, pg) == INVALID_ADDRESS)
+                return false;
+    }
+    return true;
+}
+
+void wake_robust_list(process p, void *head)
+{
+    struct robust_list_head *h = head;
+    struct robust_list *l;
+    int *uaddr;
+
+    /* must be very careful accessing the head as well as the list */
+    if (!valid_user_address(p, h, sizeof *h))
+        return;
+
+    /* XXX could keep a list of futexes and wake them at the end
+     * to let threads acquire multiple locks without blocking */
+    if (h->list_op_pending) {
+        uaddr = FUTEX_KEY_ADDR(h->list_op_pending, h->futex_offset);
+        if (valid_user_address(p, uaddr, sizeof *uaddr)) {
+            *uaddr |= FUTEX_OWNER_DIED;
+            futex_wake_many_by_uaddr(p, uaddr, 1);
+        }
+    }
+
+    for (l = h->list; (void *)l != (void *)h; l = l->next) {
+        uaddr = FUTEX_KEY_ADDR(l, h->futex_offset);
+        if (!valid_user_address(p, l, sizeof *l))
+            break;
+        if (!valid_user_address(p, uaddr, sizeof *uaddr))
+            break;
+        *uaddr |= FUTEX_OWNER_DIED;
+        futex_wake_many_by_uaddr(p, uaddr, 1);
+    }
+}
+
+sysreturn get_robust_list(int pid, void *head, u64 *len)
+{
+    struct robust_list_head **hp = head;
+    if (!validate_user_memory(hp, sizeof *hp, true))
+        return -EFAULT;
+    if (!validate_user_memory(len, sizeof *len, true))
+        return -EFAULT;
+
+    thread_log(current, "get_robust_list syscall for pid %d\n", pid);
+
+    thread t = 0;
+    if (pid == 0)
+        t = current;
+    else {
+        process p = current->p;
+        thread tt;
+        vector_foreach(p->threads, tt) {
+            if (tt->tid == pid) {
+                t = tt;
+                break;
+            }
+        }
+    }
+    if (t == 0)
+        return -ESRCH;
+    *hp = t->robust_list;
+    *len = sizeof **hp;
+    return 0;
+}
+
+sysreturn set_robust_list(void *head, u64 len)
+{
+    thread_log(current, "set_robust_list syscall with head %p", head);
+    if (len != sizeof(struct robust_list_head))
+        return -EINVAL;
+    if (!validate_user_memory(head, len, true))
```

Comment:
Should validate_user_address be called for this range as well?

Reply:
I actually tried to roll this new validation into validate_user_address but it breaks some other runtime tests that use addresses that are not in the process vmaps ranges in syscalls, so I left it separate for now. I used the new validation here instead of validate_user_address.